crates/ruff/src/rules/pylint/rules/no_self_use.rs

```diff
@@ -0,0 +1,99 @@
+use ast::Parameter;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, ParameterWithDefault};
+use ruff_python_semantic::Scope;
+
+use crate::{checkers::ast::Checker, rules::pylint::helpers::is_known_dunder_method};
+
+/// ## What it does
+/// Checks for the presence of unused `self` parameter in methods definitions.
+///
+/// ## Why is this bad?
+/// If you are not using the `self` parameter within your method, is likely
+/// that this method may not be what you want. You should consider changing
+/// to a function outside of the class or using the `@staticmethod` decorator.
+///
+/// ## Example
+/// ```python
+/// class Person:
+///     def greeting_1(self):
+///         print("Greetings friend!")
+///
+///     def greeting_2(self):
+///         print("Hello")
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class Person:
+///     name = "World"
+///
+///     @staticmethod
+///     def greeting_1():
+///         print("Greetings friend!")
+///
+///     def greeting_2(self):
+///         print(f"Hello {self.name}")
+/// ```
+#[violation]
+pub struct NoSelfUse {
+    method_name: String,
+}
+
+impl Violation for NoSelfUse {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let NoSelfUse { method_name } = self;
+        format!("Method `{method_name}` could be a function or a static method (no-self-use)")
+    }
+}
+
+/// PLR6301
+pub(crate) fn no_self_use(checker: &Checker, scope: &Scope, diagnostics: &mut Vec<Diagnostic>) {
```

Comment:
I think this should probably exclude abstract methods. We may also want to omit functions whose body is "empty" (which we typically define as just a `pass` or `...`, with or without a docstring). You can see `unused_arguments` for an example: in the method case, we omit abstract methods, overloads, empty bodies, etc.

Reply:
Ideally, we'd keep this as similar to `unused_arguments` as possible, to make it easier to maintain and so that the logic stays in-sync. We could even consider making this rule a codepath in the `unused_arguments` method?

Reply:
> Ideally, we'd keep this as similar to `unused_arguments` as possible, to make it easier to maintain and so that the logic stays in-sync. We could even consider making this rule a codepath in the `unused_arguments` method?

Yeah, it's probably better to make it a codepath in `unused_arguments`